R/try_regional_data.R

```diff
@@ -0,0 +1,125 @@
+#' @inheritParams return_data
+#' @inheritParams initialise_dataclass
+#' @return A tibble with data related to cases, deaths, hospitalisations,
+#' recoveries and testing stratified by regions within the given country.
+#' @importFrom lifecycle deprecated is_present deprecate_warn
+#' @family interface
+#' @seealso [Italy()], [UK()]
+#' @export
+#' @examples
+#' \dontrun{
+#' # set up a data cache
+#' start_using_memoise()
+#'
+#' # download data for Italy
+#' try_regional_data("italy")
+#'
+#' # return totals for Italy with no localisation
+#' try_regional_data("italy", localise = FALSE, totals = TRUE)
+#'
+#' # download data for the UK but return the class
+#' uk <- try_regional_data("United Kingdom", class = TRUE)
+#' uk
+#'
+#' # return UK data from the class object]
+#' uk$return()
+#' }
+try_regional_data <- function(country, level = "1", totals = FALSE,
+                              localise = TRUE, steps = FALSE,
+                              class = FALSE, verbose = TRUE, regions,
+                              include_level_2_regions = deprecated(),
+                              localise_regions = deprecated(),
+                              ...) {
+  if (is_present(include_level_2_regions)) {
+    deprecate_warn(
+      "0.9.0",
+      "covidregionaldata::try_regional_data(include_level_2_regions = )", "covidregionaldata::try_regional_data(level = )"
+    )
+    if (include_level_2_regions) {
+      level <- "1"
+    } else {
+      level <- "2"
+    }
+  }
+
+  if (is_present(localise_regions)) {
+    deprecate_warn(
+      "0.9.0",
+      "covidregionaldata::try_regional_data(localise_regions = )", "covidregionaldata::try_regional_data(localise = )"
+    )
+    localise <- localise_regions
+  }
+
+  # construct short hand options
+  title_country <- str_to_title(country)
+  nospace <- str_replace_all(title_country, " ", "")
+  targets <- c(
+    title_country, toupper(title_country), nospace, toupper(nospace)
+  )
+  
+  # check we have data for desired class
+  datasets <- covidregionaldata::get_available_datasets("regional")
+  target_class <- bind_rows(
+    filter(datasets, map_lgl(.data$class, ~ any(str_detect(., targets)))),
+    filter(datasets, map_lgl(.data$origin, ~ any(str_detect(., targets))))
+  ) %>%
+    distinct()
+
+  if (nrow(target_class) != 0) {
+    # check data availability and initiate country class if available
+    region_class <- initialise_dataclass(
+      class = country, level = level, regions = regions,
+      totals = totals, localise = localise,
+      verbose = verbose, steps = steps, get = TRUE,
+      type = "regional", ...
+    )
+
+    return(
+      return_data(region_class, class = class)
+    )
+  } else {
+    google_sources <- csv_reader(Google$public_fields$common_data_urls$index)
+
+    google_inputs <-
+      filter(google_sources, map_lgl(.data$country_name, ~ any(str_detect(., targets))))
+    if (nrow(google_inputs) != 0) {
+      message_verbose(verbose = verbose,
```

Comment:
Suggest moving this into the step above to make it easier to generalise this step (as could loop over Google/JHU potentially with only the code checking their supported countries need to change

Reply:
I know this was originally intended to compare over Google and JHU but JHU only appears to support 8 countries, of which we natively do a few, while Google does a stupendous number - and I didn't see an obvious way to "glance" at the JHU data without downloading several (hundred) megabytes of data which we might not use.

So this was written to just go to Google, but with the intent to avoid downloading the massive Google files if it turns out someone has asked for level 2 region data for "Furgleburf", or (more frustratingly), "Spian".

I don't use either JHU or Google data (though I might start) so I don't have a sense of the utility of writing the JHU code (which would kick in for about 4-5 countries). I'd also welcome advice on which data we may prefer to offer (where JHU and Google both offer, but we don't have specific code).

Reply:
Seems sensible to stick with just Google.